doc: add chat context documentation (#719)
eclipse-theia/theia-website: src/docs/theia_ai.md

bind(AIVariableContribution).to(TodayVariableContribution).inSingletonScope();
```

> It can now be used in any prompt template, as well as in user requests.

Reviewer: Why remove this line?

> ### Chat Context Variables

Reviewer: Should we mention how to turn this on/off (AFAIR it is possible?)

Author: Good point, I'll added it to the doc.

};

@injectable()
export class FileVariableContribution implements AIVariableContribution, AIVariableResolver {
    @inject(FileService)
    protected readonly fileService: FileService;

    @inject(WorkspaceService)
    protected readonly wsService: WorkspaceService;

    registerVariables(service: AIVariableService): void {
        service.registerResolver(FILE_VARIABLE, this);
    }

    async canResolve(request: AIVariableResolutionRequest, _: AIVariableContext): Promise<number> {
        return request.variable.name === FILE_VARIABLE.name ? 1 : 0;
    }

    async resolve(request: AIVariableResolutionRequest, _: AIVariableContext): Promise<ResolvedAIContextVariable | undefined> {
        //...
        try {
            const content = await this.fileService.readFile(path);
            return {
                variable: request.variable,
                value: await this.wsService.getWorkspaceRelativePath(path),
                contextValue: content.value.toString(),
            };
        } catch (error) {
            return undefined;
        }
    }
    //...
}
```

#### File Chat Variable Contributions

> To enhance usability, additional contributions are implemented to handle argument picking, auto-completion, and drag-and-drop support for file context variables.

Reviewer: ```suggestion
To enhance usability, an additional `FrontendVariableContribution` can be provided  to handle argument picking, auto-completion, and drag-and-drop support for file context variables.
```

    ): Promise<monaco.languages.CompletionItem[] | undefined> {
        // triggered when the user auto-completes after `#` or `#<variable-name>:`
        // your implementation needs to return a list of completion items
        //...
    }

    protected async handleDrop(event: DragEvent, _: AIVariableContext): Promise<AIVariableDropResult | undefined> {
        // triggered when the user drags an element into the chat input
        // your implementation needs to read the `event.dataTransfer` and return a variable to add to the request
        // and optionally a text to add to the chat input
        const data = event.dataTransfer?.getData('selected-tree-nodes');
        if (!data) {
            return undefined;
        }

        try {
            //...
            return { variables, text };
        } catch {
            return undefined;
        }
    }
}
```

#### Using the Context Variables from Agents

Your custom agent can now use the context variables in various ways. The context variables are available in the `context` property of the `ChatRequestModel` and apply your custom logic to decide in which way your agent passes the context data to the LLM.

> Your agents can use the variables `#contextSummary` or `#contextDetails`, which resolve to either a list or the full context, in your system message to transfer the attached context to the LLM. Alternatively, you can use the tool functions `~{context_ListChatContext}` and `~{context_ResolveChatContext}` to allow the LLM obtaining the context on demand.

Reviewer: ```suggestion
Your agents can use the predefined variables `#contextSummary` or `#contextDetails` in their prompt, which resolve to either a list or the full context, in your system message to transfer the attached context to the LLM. Alternatively, you can use the tool functions `~{context_ListChatContext}` and `~{context_ResolveChatContext}` to allow the LLM obtaining the context on demand. Please note that ignoring the context completly in you agent while displaying the context user interface in the chat will likely lead to unexpected user results.
```
